Boost: Update Cache UI to save/load settings

dilirity opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Part of  https://github.com/Automattic/jetpack/issues/35054.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* [x] Update Cache UI to save/load exclude and logging settings to/from cache config file.
* [x] Add notice when there was a problem saving excludes;
* [x] Update meta description to change based on enabled settings;
* [x] Add link to logs page when enabling logging.

There will be a follow-up to wrap-up the UI.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

p1HpG7-r9A-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

no

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Play around with the UI and make sure settings are saved/loaded correctly.

dilirity committed:
Update Cache UI to save/load logging option

dilirity committed:
Update Cache UI to save/load excludes option

dilirity committed:
add changelog

dilirity committed:
Update wording around excludes to be unified across UI and backend

dilirity committed:
Update exceptions description to closer match designs

dilirity committed:
Update description to reflect status of settings

dilirity committed:
Add link to logs when logging is enabled

dilirity committed:
Try to show error notice if something's wrong when saving exceptions

dilirity committed:
Update Cache UI to always be collapsed

dilirity committed:
Fix not showing error notice for exceptions

dilirity committed:
Update exceptions UI to show an error state

dilirity committed:
Merge branch 'trunk' into update/boost/cache-ui-save-load-settings

dilirity committed:
Attempt to fix build

pyronaur commented on projects/plugins/boost/app/assets/src/js/features/page-cache/meta/meta.tsx:
```diff
@@ -1,19 +1,69 @@
+						__( 'No exceptions or logging.', 'jetpack-boost' )
+					) : (
+						<>
+							{ totalExceptions > 0 ? (
```
We're checking for `totalExceptions` equality twice, that results in both messages like `No Exceptions or logging` and `No exceptions`. Can the logic be changed so that we display more precise messages, like `No exceptions` and `Logging isn't enabled`?

dilirity replied:
I'm not sure what you mean here. I don't see how `No exceptions or logging.` and `No exceptions.` would show up at the same time. Did you actually manage to see this?

Though the whole logic could use a tune up. It's kind of messy.

dilirity replied:
As discussed over Slack, I'm leaving this for now.

pyronaur commented on projects/plugins/boost/app/assets/src/js/features/page-cache/meta/meta.tsx:
```diff
@@ -1,19 +1,69 @@
 
 const Meta = () => {
 	const [ isExpanded, setIsExpanded ] = useState( false );
+	const [ query, mutation ] = usePageCache();
```
Looks like `usePageCache` might be using DataSync. Should this be wrapped in an Error Boundary? ( https://wp.me/pc9hqz-2wT ).

dilirity replied:
Thanks! Updated.

pyronaur commented on projects/plugins/boost/app/assets/src/js/features/page-cache/meta/meta.tsx:
```diff
@@ -39,29 +89,121 @@ const Meta = () => {
+	function showExample( event: React.MouseEvent ) {
+		event.preventDefault();
+
+		// @todo - add proper example.
```
I think the functionality can be probably implemented with the text being "Lorem ipsum...."

dilirity replied:
I've removed the functionality for the time being. We'll implement it later.

pyronaur commented on projects/plugins/boost/app/assets/src/js/features/page-cache/meta/meta.tsx:
```diff
@@ -39,29 +89,121 @@ const Meta = () => {
+	const [ inputInvalid, setInputInvalid ] = useState( false ); // @todo - implement this
+
+	// @todo - add proper link.
+	const exclusionsLink = 'TBD';
```
Use `https://jetpack.com` for now so that the link renders corretly.

dilirity replied:
Thanks. Updated.

pyronaur reviewed (changes requested):
![2024_02_20bTFFNV9R 04](https://github.com/Automattic/jetpack/assets/988095/0be2e73f-bb6f-4911-9ad8-6335a6821976)

Without enabling logging - on the first load I already have a log, is this expected?

dilirity committed:
Remove WIP code

dilirity committed:
Update link

dilirity committed:
Fix showing settings UI when there's an error

dilirity committed:
Update wording around cache excludes to make more sense

dilirity committed:
Add error boundary for PageCache component

dilirity commented:
> Without enabling logging - on the first load I already have a log, is this expected?

It is currently expected as this PR only stores settings in the config. It's up to logging to respect the config. Will be addressed in another PR.

I've also fixed a bug shown from the screenshot you provided, where the Settings were showing above the Error notice. If there's an error, only the Error notice will now show.

I've also updated the naming around "exceptions" in code to make more sense.

pyronaur commented on projects/plugins/boost/app/modules/cache/data-sync/Page_Cache_Entry.php:
```diff
@@ -0,0 +1,45 @@
+		$cache_settings = Boost_Cache_Settings::get_instance();
+
+		$settings = array(
+			'bypassPatterns' => $cache_settings->get_bypass_patterns(),
```
DataSync keys (any JSON keys) should be `snake_case` for consistency. Otherwise we have `window.jetpack_boost_ds.page_cache.bypassPatterns`

dilirity replied:
Thanks!

Fixed in https://github.com/Automattic/jetpack/pull/35729/commits/073486060406b0b59dfe8c557ef2198b423f1141.

pyronaur commented on projects/plugins/boost/app/modules/cache/pre-wordpress/Request.php:
```diff
@@ -83,11 +83,11 @@ public function is_url_excluded( $request_uri = '' ) {
-		$excluded_urls = Boost_Cache_Settings::get_instance()->get_excluded_urls();
-		$excluded_urls = apply_filters( 'boost_cache_excluded_urls', $excluded_urls );
+		$exceptions = Boost_Cache_Settings::get_instance()->get_bypass_patterns();
+		$exceptions = apply_filters( 'boost_cache_exceptions', $exceptions );
```
`boost_cache_bypass_patterns` ?

dilirity replied:
Updated in 7b8da7d.

Makes more sense!

dilirity committed:
Update variable name

dilirity committed:
Update filter name

pyronaur reviewed (approved):
Great work! 👏
